### PR TITLE
feat(ast): check_ast func and test cases

### PR DESCRIPTION
**Pull request by yuanbohan**

### CODE REVIEW

File: src/label/matcher.rs [hunk 1]
```diff
@@ -126,6 +126,24 @@ impl Matchers {
...
+    /// matcher that does not match the empty string.
+    ///
+    /// The following expression is illegal:
+    /// {job=~".*"} # Bad!
+    pub fn is_empty_matchers(&self) -> bool {
+        self.matchers.is_empty() || self.matchers.iter().all(|m| m.is_match(""))
+    }
+
+    /// check if the specified name exists more than once in matchers.
+    pub fn duplicated_matchers(&self, name: &str) -> Vec<&String> {
```
Comment (evenyag) [inline src/label/matcher.rs:140]: Looks like this method doesn't do the check, it just returns all matchers with specific name.

File: src/parser/ast.rs [hunk 2]
```diff
@@ -789,6 +817,132 @@ fn check_ast_for_binary_expr(mut ex: BinaryExpr) -> Result<Expr, String> {
...
+
+    let expect_param_type =
+        |actual: Option<&Expr>, expected: ValueType, context: &str| -> Result<bool, String> {
+            match actual {
+                Some(expr) => expect_type(Some(expr.value_type()), expected, context),
+                None => expect_type(None, expected, context),
+            }
+        };
+
+    if ex.op.id() == T_TOPK || ex.op.id() == T_BOTTOMK || ex.op.id() == T_QUANTILE {
```
Comment (evenyag) [inline src/parser/ast.rs:842]:
```suggestion
    if matches!(ex.op.id(), T_TOPK | T_BOTTOMK | T_QUANTILE) {
```

File: src/parser/ast.rs [hunk 3]
```diff
@@ -789,6 +817,132 @@ fn check_ast_for_binary_expr(mut ex: BinaryExpr) -> Result<Expr, String> {
...
+        expect_param_type(
+            ex.param.as_deref(),
```
Comment (evenyag) [inline src/parser/ast.rs:844]: Could we use `as_ref()`? Seems `ex.param` is just `Option<Expr>`, not a string or pointer type.

File: src/parser/ast.rs [hunk 4]
```diff
@@ -789,6 +817,132 @@ fn check_ast_for_binary_expr(mut ex: BinaryExpr) -> Result<Expr, String> {
...
+
+fn check_ast_for_vector_selector(ex: VectorSelector) -> Result<Expr, String> {
+    // A Vector selector must contain at least one non-empty matcher to prevent
+    // implicit selection of all metrics (e.g. by a typo).
+    if ex.matchers.is_empty_matchers() {
+        return Err("vector selector must contain at least one non-empty matcher".into());
+    }
+
+    let mut du = ex.matchers.duplicated_matchers(METRIC_NAME);
+    du.sort();
```
Comment (evenyag) [inline src/parser/ast.rs:934]: Is it necessary to sort the `du`?

File: src/parser/ast.rs [hunk 5]
```diff
@@ -789,6 +817,132 @@ fn check_ast_for_binary_expr(mut ex: BinaryExpr) -> Result<Expr, String> {
...
+    if du.len() >= 2 {
+        let du1 = du[0];
+        let du2 = du[1];
+        return Err(format!(
+            "metric name must not be set twice: '{du1}' or '{du2}'"
+        ));
```
Comment (evenyag) [inline src/parser/ast.rs:940]:
```suggestion
        return Err(format!(
            "metric name must not be set twice: '{}' or '{}'", du[0], du[1]
        ));
```

**Review by evenyag — COMMENTED** [review]
I left some comments about style.

File: src/label/matcher.rs (hunk 1 above)
Comment (yuanbohan) [reply src/label/matcher.rs:140]: good catch, the docstring is misleading.

File: src/parser/ast.rs (hunk 3 above)
Comment (yuanbohan) [reply src/parser/ast.rs:844]: it's an `Option<Box<Expr>>`, and compiler helps me to use `as_deref`

File: src/parser/ast.rs (hunk 4 above)
Comment (yuanbohan) [reply src/parser/ast.rs]: actually, it depends. For PromQL user, no. but in test cases, we have to compare the string literal, and HashSet does not guarantee the order, so the the order in `du` is not guaranteed either.

In order not to fail the test cases occasionally, I use the `sort` method.

File: src/parser/ast.rs (hunk 3 above)
Comment (evenyag) [reply src/parser/ast.rs:844]: Looks good. Seems the compiler can auto-deref it for you.

Comment (yuanbohan) [reply src/parser/ast.rs]: I have deleted the `expect_param_type`, and directly use the `expect_type` function

**Review by evenyag — APPROVED** [review]

**Pull request merged**